src/Foundatio/Utility/Time.cs

```diff
@@ -0,0 +1,224 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Foundatio.Utility {
+    public interface ITime {
+        DateTime Now();
+        DateTime UtcNow();
+        DateTimeOffset OffsetNow();
+        DateTimeOffset OffsetUtcNow();
+        TimeSpan TimeZoneOffset();
+        void Delay(int milliseconds);
+        Task DelayAsync(int milliseconds, CancellationToken ct);
+    }
+
+    public class RealTime : ITime {
+        public static readonly RealTime Instance = new RealTime();
+
+        public DateTime Now() => DateTime.Now;
+        public DateTime UtcNow() => DateTime.UtcNow;
+        public DateTimeOffset OffsetNow() => DateTimeOffset.Now;
+        public DateTimeOffset OffsetUtcNow() => DateTimeOffset.UtcNow;
+        public TimeSpan TimeZoneOffset() => DateTimeOffset.Now.Offset;
+        public void Delay(int milliseconds) => Thread.Sleep(milliseconds);
+        public Task DelayAsync(int milliseconds, CancellationToken ct) => Task.Delay(milliseconds, ct);
+    }
+
+    public class TestTime : ITime, IDisposable {
+        private DateTime? _fixedUtc = null;
+        private TimeSpan _offset = TimeSpan.Zero;
+        private TimeSpan _timeZoneOffset = DateTimeOffset.Now.Offset;
+        private bool _fakeSleep = false;
+        
+        public TestTime() {}
+
+        public TestTime(ITime originalTime) {
+            _originalTime = originalTime;
+        }
+
+        public DateTime UtcNow() => (_fixedUtc ?? DateTime.UtcNow).Add(_offset);
+        public DateTime Now() => new DateTime(UtcNow().Ticks + TimeZoneOffset().Ticks, DateTimeKind.Local);
+        public DateTimeOffset OffsetNow() => new DateTimeOffset(UtcNow().Ticks + TimeZoneOffset().Ticks, TimeZoneOffset());
+        public DateTimeOffset OffsetUtcNow() => new DateTimeOffset(UtcNow().Ticks, TimeSpan.Zero);
+        public void SetTimeZoneOffset(TimeSpan offset) => _timeZoneOffset = offset;
+        public void AddTime(TimeSpan amount) => _offset = _offset.Add(amount);
+        public void SubtractTime(TimeSpan amount) => _offset = _offset.Subtract(amount);
+        public void UseFakeSleep() => _fakeSleep = true;
+        public void UseRealSleep() => _fakeSleep = false;
+
+        public void Delay(int milliseconds) {
+            if (!_fakeSleep) {
+                Thread.Sleep(milliseconds);
+                return;
+            }
+
+            AddTime(TimeSpan.FromMilliseconds(milliseconds));
+            Thread.Sleep(1);
+        }
+
+        public Task DelayAsync(int milliseconds, CancellationToken ct) {
+            if (!_fakeSleep)
+                return Task.Delay(milliseconds, ct);
+
+            Delay(milliseconds);
+            return Task.CompletedTask;
+        }
+
+        public TimeSpan TimeZoneOffset() => _timeZoneOffset;
+
+        public void Freeze() {
+            SetFrozenTime(Now());
+        }
+
+        public void Unfreeze() {
+            SetTime(Now());
+        }
+
+        public void SetFrozenTime(DateTime time) {
```

Comment:
Could be simplified into `public void Freeze(DateTime time) {`, making use of `Freeze` for both scenarious.